File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        //define 3 variables: prev, curr, next
        //prev null, curr head, next, head.next
        //while curr is not null
        //curr.next to prev
        //if next is null
        //break <- reached the end of the linked list and curr points to the new head
        //prev is now curr
        //curr is now next
        //next is now next.next

        if(head == null)
            return head;

        ListNode prev = null;
        ListNode curr = head;
        ListNode next;

        while(curr != null)
        {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # have a loop that runs as long as both lists still has a node in it
        # create a head node (dummy node) append nodes to the end of this
        # at each iteration, compare list1 and list2
        # check that either list still has a node in it, append that whole list to the end of the linked list
        # return head.next

        head = cur = ListNode()

        while list1 and list2:
            if list1.val < list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next
            cur = cur.next
        
        if list1:
            cur.next = list1
        else:
            cur.next = list2

        return head.next